File: php/src/VedaMetrics.php
<?php

declare(strict_types=1);

namespace VedaDB;

/**
 * Prometheus-compatible metrics collector for VedaDB operations.
 */
class VedaMetrics
{
    private string $namespace;

    /** @var array<string, array{value: float, labels: array<string, string>, timestamp: float}[]> */
    private array $gauges = [];

    /** @var array<string, array{count: float, sum: float, labels: array<string, string>}> */
    private array $histograms = [];

    /** @var array<string, float> */
    private array $counters = [];

    private float $startTime;

    /** @var array<string, array{count: int, total_time: float, errors: int}> */
    private array $queryStats = [];

    public function __construct(string $namespace = 'vedadb')
    {
        $this->namespace = $namespace;
        $this->startTime = microtime(true);
    }

    /**
     * Record a query execution.
     */
    public function recordQuery(string $operation, float $durationMs, bool $error = false, array $labels = []): void
    {
        $key = $operation;
        if (!isset($this->queryStats[$key])) {
            $this->queryStats[$key] = [
                'count'      => 0,
                'total_time' => 0.0,
                'errors'     => 0,
            ];
        }

        $this->queryStats[$key]['count']++;
        $this->queryStats[$key]['total_time'] += $durationMs;
        if ($error) {
            $this->queryStats[$key]['errors']++;
        }

        // Counter for total queries
        $counterKey = "{$this->namespace}_queries_total";
        $this->counters[$counterKey] = ($this->counters[$counterKey] ?? 0) + 1;

        // Histogram for duration
        $this->recordHistogram('query_duration_ms', $durationMs, array_merge(
            ['operation' => $operation],
            $labels,
        ));
    }

    /**
     * Record a gauge value.
     */
    public function recordGauge(string $name, float $value, array $labels = []): void
    {
        $key = "{$this->namespace}_{$name}";
        if (!isset($this->gauges[$key])) {
            $this->gauges[$key] = [];
        }

        $labelKey = $this->labelsKey($labels);
        $this->gauges[$key][$labelKey] = [
            'value'     => $value,
            'labels'    => $labels,
            'timestamp' => microtime(true),
        ];
    }

    /**
     * Record a histogram observation.
     */
    public function recordHistogram(string $name, float $value, array $labels = []): void
    {
        $key = "{$this->namespace}_{$name}";
        $labelKey = $this->labelsKey($labels);

        if (!isset($this->histograms[$key][$labelKey])) {
            $this->histograms[$key][$labelKey] = [
                'count'  => 0.0,
                'sum'    => 0.0,
                'labels' => $labels,
            ];
        }

        $this->histograms[$key][$labelKey]['count']++;
        $this->histograms[$key][$labelKey]['sum'] += $value;
    }

    /**
     * Increment a counter.
     */
    public function incrementCounter(string $name, float $amount = 1.0, array $labels = []): void
    {
        $key = "{$this->namespace}_{$name}";
        $labelKey = $this->labelsKey($labels);
        $fullKey = $key . '_' . $labelKey;
        $this->counters[$fullKey] = ($this->counters[$fullKey] ?? 0) + $amount;
    }

    /**
     * Record connection pool metrics.
     */
    public function recordPoolMetrics(int $active, int $idle, int $maxSize): void
    {
        $this->recordGauge('pool_active_connections', (float) $active);
        $this->recordGauge('pool_idle_connections', (float) $idle);
        $this->recordGauge('pool_max_connections', (float) $maxSize);
    }

    /**
     * Record circuit breaker state.
     */
    public function recordCircuitBreaker(string $name, string $state): void
    {
        $this->recordGauge('circuit_breaker_state', match ($state) {
            'closed'     => 0.0,
            'half_open'  => 1.0,
            'open'       => 2.0,
            default      => 3.0,
        }, ['name' => $name]);
    }

    /**
     * Record cache metrics.
     */
    public function recordCacheMetrics(int $hits, int $misses, int $size): void
    {
        $this->recordGauge('cache_hits_total', (float) $hits);
        $this->recordGauge('cache_misses_total', (float) $misses);
        $this->recordGauge('cache_size', (float) $size);

        $total = $hits + $misses;
        if ($total > 0) {
            $this->recordGauge('cache_hit_ratio', round($hits / $total, 4));
        }
    }

    /**
     * Export metrics in Prometheus text format.
     */
    public function exportPrometheus(): string
    {
        $lines = [];
        $lines[] = "# VedaDB PHP Driver Metrics";
        $lines[] = "# HELP {$this->namespace}_info Driver information";
        $lines[] = "# TYPE {$this->namespace}_info gauge";
        $lines[] = "{$this->namespace}_info{version=\"1.0.0\",language=\"php\"} 1";
        $lines[] = "";

        // Counters
        foreach ($this->counters as $key => $value) {
            $parts = explode('_', $key, 2);
            $metricName = $parts[0] . '_' . ($parts[1] ?? 'unknown');
            $lines[] = "# TYPE {$metricName} counter";
            $lines[] = "{$metricName} {$value}";
        }

        // Gauges
        foreach ($this->gauges as $name => $entries) {
            $lines[] = "# TYPE {$name} gauge";
            foreach ($entries as $entry) {
                $labelStr = $this->formatLabels($entry['labels']);
                $lines[] = "{$name}{$labelStr} {$entry['value']}";
            }
        }

        // Histograms
        foreach ($this->histograms as $name => $entries) {
            $lines[] = "# TYPE {$name} summary";
            foreach ($entries as $entry) {
                $labelStr = $this->formatLabels($entry['labels']);
                $lines[] = "{$name}_count{$labelStr} {$entry['count']}";
                $lines[] = "{$name}_sum{$labelStr} {$entry['sum']}";
            }
        }

        // Query stats
        foreach ($this->queryStats as $op => $stats) {
            $labelStr = "{operation=\"{$op}\"}";
            $lines[] = "# TYPE {$this->namespace}_query_total counter";
            $lines[] = "{$this->namespace}_query_total{$labelStr} {$stats['count']}";
            if ($stats['count'] > 0) {
                $avg = round($stats['total_time'] / $stats['count'], 3);
                $lines[] = "{$this->namespace}_query_duration_avg{$labelStr} {$avg}";
            }
            $lines[] = "{$this->namespace}_query_errors_total{$labelStr} {$stats['errors']}";
        }

        return implode("\n", $lines) . "\n";
    }

    /**
     * Get metrics as an array.
     */
    public function getMetrics(): array
    {
        return [
            'uptime_seconds'  => round(microtime(true) - $this->startTime, 3),
            'query_stats'     => $this->queryStats,
            'gauges'          => $this->gauges,
            'histograms'      => $this->histograms,
            'counters'        => $this->counters,
        ];
    }

    /**
     * Reset all metrics.
     */
    public function reset(): void
    {
        $this->gauges     = [];
        $this->histograms = [];
        $this->counters   = [];
        $this->queryStats = [];
    }

    private function labelsKey(array $labels): string
    {
        if (empty($labels)) {
            return '_default';
        }
        ksort($labels);
        return md5(implode(',', array_map(
            fn(string $k, string $v) => "{$k}={$v}",
            array_keys($labels),
            array_values($labels),
        )));
    }

    private function formatLabels(array $labels): string
    {
        if (empty($labels)) {
            return '';
        }
        $pairs = [];
        foreach ($labels as $k => $v) {
            $pairs[] = "{$k}=\"{$v}\"";
        }
        return '{' . implode(',', $pairs) . '}';
    }
}

File: dotnet/VedaMetrics.cs
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Text;
using System;

namespace VedaDB;

/// <summary>
/// Metrics collector for VedaDB driver operations.
/// Emits Prometheus-compatible text format for scraping.
/// </summary>
public static class VedaMetrics
{
    private static readonly ConcurrentDictionary<string, long> _counters = new();
    private static readonly ConcurrentDictionary<string, double> _gauges = new();
    private static readonly ConcurrentDictionary<string, List<double>> _histograms = new();
    private static readonly ConcurrentDictionary<string, Stopwatch> _timers = new();

    /// <summary>
    /// Whether metrics collection is enabled.
    /// </summary>
    public static bool Enabled { get; set; } = true;

    /// <summary>
    /// Increment a counter metric.
    /// </summary>
    public static void Increment(string name, long value = 1, Dictionary<string, string>? labels = null)
    {
        if (!Enabled) return;
        var key = FormatKey(name, labels);
        _counters.AddOrUpdate(key, value, (_, existing) => existing + value);
    }

    /// <summary>
    /// Set a gauge metric.
    /// </summary>
    public static void Gauge(string name, double value, Dictionary<string, string>? labels = null)
    {
        if (!Enabled) return;
        var key = FormatKey(name, labels);
        _gauges[key] = value;
    }

    /// <summary>
    /// Record a value in a histogram.
    /// </summary>
    public static void Histogram(string name, double value, Dictionary<string, string>? labels = null)
    {
        if (!Enabled) return;
        var key = FormatKey(name, labels);
        _histograms.AddOrUpdate(key, _ => new List<double> { value }, (_, list) =>
        {
            lock (list) { list.Add(value); }
            return list;
        });
    }

    /// <summary>
    /// Start a timer for an operation.
    /// </summary>
    public static void TimerStart(string name)
    {
        if (!Enabled) return;
        _timers[name] = Stopwatch.StartNew();
    }

    /// <summary>
    /// Stop a timer and record the duration in seconds.
    /// </summary>
    public static double TimerStop(string name)
    {
        if (!Enabled || !_timers.TryRemove(name, out var sw)) return 0;
        sw.Stop();
        var seconds = sw.Elapsed.TotalSeconds;
        Histogram($"{name}_duration_seconds", seconds);
        return seconds;
    }

    /// <summary>
    /// Get all counter values.
    /// </summary>
    public static IReadOnlyDictionary<string, long> Counters => _counters;

    /// <summary>
    /// Get all gauge values.
    /// </summary>
    public static IReadOnlyDictionary<string, double> Gauges => _gauges;

    /// <summary>
    /// Reset all metrics.
    /// </summary>
    public static void Reset()
    {
        _counters.Clear();
        _gauges.Clear();
        _histograms.Clear();
        _timers.Clear();
    }

    /// <summary>
    /// Export metrics in Prometheus text format.
    /// </summary>
    public static string ExportPrometheus()
    {
        var sb = new StringBuilder();

        foreach (var (key, value) in _counters)
        {
            var (name, labels) = ParseKey(key);
            sb.AppendLine($"# TYPE {name} counter");
            if (labels != null)
                sb.AppendLine($"{name}{{{labels}}} {value}");
            else
                sb.AppendLine($"{name} {value}");
        }

        foreach (var (key, value) in _gauges)
        {
            var (name, labels) = ParseKey(key);
            sb.AppendLine($"# TYPE {name} gauge");
            if (labels != null)
                sb.AppendLine($"{name}{{{labels}}} {value}");
            else
                sb.AppendLine($"{name} {value}");
        }

        foreach (var (key, values) in _histograms)
        {
            var (name, labels) = ParseKey(key);
            sb.AppendLine($"# TYPE {name} histogram");
            double[] snapshot;
            lock (values) { snapshot = values.ToArray(); }

            var buckets = new[] { 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0 };
            foreach (var bucket in buckets)
            {
                var count = snapshot.Count(v => v <= bucket);
                var bucketLabel = labels != null
                    ? $"le=\"{bucket}\",{labels.TrimStart('}').TrimEnd('}')}"
                    : $"le=\"{bucket}\"";
                sb.AppendLine($"{name}_bucket{{{bucketLabel}}} {count}");
            }
            sb.AppendLine($"{name}_count {snapshot.Length}");
            if (snapshot.Length > 0)
                sb.AppendLine($"{name}_sum {snapshot.Sum()}");
        }

        return sb.ToString();
    }

    private static string FormatKey(string name, Dictionary<string, string>? labels)
    {
        if (labels == null || labels.Count == 0) return name;
        var labelStr = string.Join(",", labels.Select(kv => $"{kv.Key}=\"{kv.Value}\""));
        return $"{name}{{{labelStr}}}";
    }

    private static (string Name, string? Labels) ParseKey(string key)
    {
        var braceIdx = key.IndexOf('{');
        if (braceIdx < 0) return (key, null);
        return (key.Substring(0, braceIdx), key.Substring(braceIdx + 1, key.Length - braceIdx - 2));
    }
}
